### PR TITLE
Fix #2894

### LINKED ISSUE #2894: `use_self` lint false positives
Minimal example:
```rust
trait IntoBytes {
    fn into_bytes(&self) -> Vec<u8>;
}

impl IntoBytes for u8 {
    fn into_bytes(&self) -> Vec<u8> {
        vec![*self]
    }
}
```
Here `u8` in the trait definition is used as `byte`, but Clippy suggest to replace `u8` with `Self` which is confusing. I think that this lint shouldn't be triggered for trait implementation.

### PR DESCRIPTION
**Pull request by mikerite**

### CODE REVIEW

**Comment by mikerite** [comment]
I broke something because the code below no longer lints on the `rhs` type or the return.

```
use std::ops::Mul;

fn main() {}

struct A;

impl Mul<A> for A {
    type Output = Self;
    
    fn mul(self, rhs: A) -> A {
        rhs
    }
}
```

File: clippy_lints/src/use_self.rs [hunk 1]
```diff
@@ -49,13 +51,93 @@ impl LintPass for UseSelf {
 
 const SEGMENTS_MSG: &str = "segments should be composed of at least 1 element";
 
+fn span_use_self_lint(cx: &LateContext, path: &Path) {
```
Comment (flip1995) [inline clippy_lints/src/use_self.rs]: This function does the same as 
https://github.com/rust-lang-nursery/rust-clippy/blob/06d6710147742892efc17ed307b3397bfe093bc7/clippy_lints/src/utils/mod.rs#L624-L631

**Comment by mikerite** [comment]
I think I've fixed it but I'm not 100% convinced of my logic. :thinking: 
I'll incorporate review suggestions later.

**Comment by mikerite** [comment]
I tried it against a few popular crates and it seems fine.

**Comment by mikerite** [comment]
@flip1995 Are you OK merging this?

**Comment by flip1995** [comment]
Didn't get to give it a closer look. I Will do it today, and get back to you. Sorry for the delay!

**Comment by flip1995** [comment]
NIT: You can use `git rebase` instead of `git merge` to avoid this merge commits.

**Comment by mikerite** [comment]
Force pushing to a PR breaks the 'View changes' links so I try to avoid that. Couldn't you have avoided the merge commits by squash merging? I'm not sure why no one seems to use that option.

**Comment by flip1995** [comment]
As long as it is just a rebase, without squashing any commits, the latest changes can also be viewed, by selecting the relevant commits, which is just a few clicks more than 'View changes', which I personally don't mind. 

Squash merging produces a single commit. I don't know how this interacts with git blame or similar... This could get pretty annoying, I could imagine.

Maybe it is just personal preference, but I adapt to the standard merge method used in this repo.

**Pull request merged**